### PR TITLE
Adds followup post on Jenkins automation

### PR DESCRIPTION
**Pull request by djablonski**

### CODE REVIEW

File: _posts/2015-10-20-jenkins-followup.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
+---
+layout: post
+title: "Followup: Automating Jenkins"
```
Comment (BirgitBader) [inline _posts/2015-10-20-jenkins-followup.md]: Follow-up

File: _posts/2015-10-20-jenkins-followup.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+date: "2015-10-20 09:00:00"
```
Comment (BirgitBader) [inline _posts/2015-10-20-jenkins-followup.md]: 2015-10-22

File: _posts/2015-10-20-jenkins-followup.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+icon: wrench
+categories: tech-stories
+authors: ["Dirk Jablonski"]
+---
+
+Some weeks ago, we wrote on [Infrstructure as Code: automating Jenkins](https://developer.epages.com/blog/2015/06/25/infrastructure-as-code.html).
```
Comment (BirgitBader) [inline _posts/2015-10-20-jenkins-followup.md]: ... we published [Infrastructure as Code: automating Jenkins]

File: _posts/2015-10-20-jenkins-followup.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+In that post, we described some open ends, which still needed to be solved in order to satisfy our expectations.
+In this followup post, I will explain how we advanced on our way to have a fully automated continuous integration server, this time coming pretty close to an *[immutable server](http://martinfowler.com/bliki/ImmutableServer.html)*.
```
Comment (BirgitBader) [inline _posts/2015-10-20-jenkins-followup.md]: In this follow-up post...

File: _posts/2015-10-20-jenkins-followup.md [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+## Remaining shortcomings
+
+The main problem we had earlier was the central configuration of the Jenkins master, which is extended by lots of plugins.
```
Comment (BirgitBader) [inline _posts/2015-10-20-jenkins-followup.md]: The main challenge...

File: _posts/2015-10-20-jenkins-followup.md [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+As a workaround, we created the full `config.xml` in advance and just copied it over to Jenkins.
+In addition, we were pretty unhappy with not being able to avoid local modifications, which were still leaving permanent traces behind, even when being overwritten by a new run of Ansible.
```
Comment (BirgitBader) [inline _posts/2015-10-20-jenkins-followup.md]: Maybe link [Ansible] to the respective website?

File: _posts/2015-10-20-jenkins-followup.md [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+## Dockerize!
```
Comment (BirgitBader) [inline _posts/2015-10-20-jenkins-followup.md]: Dockerise!

File: _posts/2015-10-20-jenkins-followup.md [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+{% image docker-logo-h.png %}{% endimage %}
+
+So, let's get started on our journey towards immutable Jenkins master!
+The main step to achieve this goal was our decision to build our further steps on [Docker](https://www.docker.com/).
```
Comment (BirgitBader) [inline _posts/2015-10-20-jenkins-followup.md]: The main step to achieve this goal was based on the decision to build further steps on...

File: _posts/2015-10-20-jenkins-followup.md [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+With the very sophisticated official [Jenkins image](https://hub.docker.com/_/jenkins/) available at [Docker Hub](https://hub.docker.com/), we found an excellent starting point.
+The image already provides out-of-the-box support for installing additional plugins by just providing a list of plugins and versions,
```
Comment (BirgitBader) [inline _posts/2015-10-20-jenkins-followup.md]: not sure if there's a redundant whitespace behind "versions,...."

File: _posts/2015-10-20-jenkins-followup.md [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+which we previously did in a much more complicated way using Jenkins CLI.
```
Comment (BirgitBader) [inline _posts/2015-10-20-jenkins-followup.md]: Maybe link [Jenkins CLI] to the respective website?

Comment (djablonski) [inline _posts/2015-10-20-jenkins-followup.md]: Is just a part of Jenkins. No need for this (this time)...

File: _posts/2015-10-20-jenkins-followup.md [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
+In addition, the Docker image hinted us to the solution of the configuration problem: Groovy init hooks.
```
Comment (BirgitBader) [inline _posts/2015-10-20-jenkins-followup.md]: what do you mean by "hinted us"? maybe "brought us"?

File: _posts/2015-10-20-jenkins-followup.md [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+With these scripts, which utilize the internal API of Jenkins itself and also the plugins, we were able to kick out the precreated `config.xml`.
```
Comment (BirgitBader) [inline _posts/2015-10-20-jenkins-followup.md]: utilise

File: _posts/2015-10-20-jenkins-followup.md [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+these scripts through templates from Ansible.
+
+## Immutability
+
+To get a truly immutable server, there are two things left: Once again, the handling of job metadata, which needs to be persisted, and the Jenkins Web UI, which still allows changing the configuration,
+although these cahnges will be overwritten with the next run of Ansible.
+
+The job metadata problem is easily solved using a standard Docker feature: volumes.
+We are intentionally not mounting the whole `jenkins_home` folder as a persistent volume, but only the reallly important subfolders `jobs` and `nodes`, which contain the actual metadata.
+Like described before, we only backup a small fraction of the files in these folders, as for example the `jobs` folder also contains the workspace,
```
Comment (BirgitBader) [inline _posts/2015-10-20-jenkins-followup.md]: As already described, ...

File: _posts/2015-10-20-jenkins-followup.md (hunk 11 above)
Comment (djablonski) [inline _posts/2015-10-20-jenkins-followup.md]: No. More like "gave us a hint in the right direction"

File: _posts/2015-10-20-jenkins-followup.md [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+which can be created whenever needed by just checking out from version control.
+
+The problem of possible modifications through the web UI can be solved using the [Matrix Authorization Strategy Plugin](https://wiki.jenkins-ci.org/display/JENKINS/Matrix+Authorization+Strategy+Plugin).
+Utilising this plugin, we create only one real admin user, which has a (unknown) random password, and other users, which are only allowed to trigger builds.
+This way, the only way to change the configuration of the server is to create a new version of the Docker container with Ansible.
```
Comment (BirgitBader) [inline _posts/2015-10-20-jenkins-followup.md]: In doing so, the only way to ...
(avoids double "way" :-))

File: _posts/2015-10-20-jenkins-followup.md [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+Also, jobs can also only be created or modified by the [Job DSL Plugin](https://wiki.jenkins-ci.org/display/JENKINS/Job+DSL+Plugin).
```
Comment (BirgitBader) [inline _posts/2015-10-20-jenkins-followup.md]: Furthermore, jobs can also be created...
(avoids double "also" :-))

File: _posts/2015-10-20-jenkins-followup.md [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+
+# Outlook
+
+Regarding the general setup of the Jenkins master, we are now more or less where we want to be.
+There are always improvements to make, but the main challanges that lie ahead of us are more in the area of getting our build slaves automated the way we want,
```
Comment (BirgitBader) [inline _posts/2015-10-20-jenkins-followup.md]: There's still room for improvement, but the main challenges...

File: _posts/2015-10-20-jenkins-followup.md [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+and running all of our build jobs in an appropriate environment.
+In the end, this is typical everyday CI maintenance, which we wont be able to get rid of completely even with full automation... ;-)
```
Comment (BirgitBader) [inline _posts/2015-10-20-jenkins-followup.md]: ... which we won't be able to...

Comment (BirgitBader) [inline _posts/2015-10-20-jenkins-followup.md]: Unfortunately, I'm not sure if we are allowed to use the Docker image.

**Comment by BirgitBader** [comment]
Thanks for this great post!
I'm done. Would you like to include the comments first or shall we continue the review?

File: _posts/2015-10-20-jenkins-followup.md (hunk 17 above)
Comment (djablonski) [inline _posts/2015-10-20-jenkins-followup.md]: For internal reasons? The usage of the image is aligned with their legal terms (see https://www.docker.com/brand-guidelines), and it is taken from their own media kit...

Comment (BirgitBader) [inline _posts/2015-10-20-jenkins-followup.md]: Perfect! Then, keep it!

**Comment by BirgitBader** [comment]
@maletzke @steffimeier please give your feedback

**Comment by BirgitBader** [comment]
@JessTinker please proof-read, thank you.

File: _posts/2015-10-20-jenkins-followup.md [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+In addition, we were pretty unhappy with not being able to avoid local modifications, which were still leaving permanent traces behind, even when being overwritten by a new run of [Ansible](http://www.ansible.com/).
+
+## Introducing Docker
+
+{% image docker-logo-h.png %}{% endimage %}
+
+So, let's get started on our journey towards immutable Jenkins master!
+The main step to achieve this goal was our decision to build our further steps on [Docker](https://www.docker.com/).
+With the very sophisticated official [Jenkins image](https://hub.docker.com/_/jenkins/) available at [Docker Hub](https://hub.docker.com/), we found an excellent starting point.
+The image already provides out-of-the-box support for installing additional plugins by just providing a list of plugins and versions, which we previously did in a much more complicated way using Jenkins CLI.
```
Comment (JessTinker) [inline _posts/2015-10-20-jenkins-followup.md]: I would remove the word just

**Comment by BirgitBader** [comment]
@djablonski Jess is done with the proof-read. Please squash the commits into one. I will merge it then and publish the post tomorrow. Thanks.

**Pull request closed without merging**